### PR TITLE
Simplesort

### PR DESCRIPTION
**Pull request by colby-anderson**
This pull request is for an unstable out-of-place single threaded sorting algorithm that supports all python types.

In the process of adding support for tuples/dicts/lists; this support should be done by the end of the day, just had a few questions that I will ask over Slack. However, a code review could start now as adding in this support requires just a few extra conditional branches. If you decide the code is too unreadable and would like me to organize it better and/or add comments, let me know and I can do that before the code review.

An unstable, in-place, multi-threaded sorting algorithm is being developed concurrently on a separate branch. This other branch uses Quicksort and radix sort.

### CODE REVIEW

File: tuplex/core/include/DataSet.h [hunk 1]
```diff
@@ -86,6 +86,16 @@ namespace tuplex {
 
         virtual ~DataSet();
 
+        /*!
+         * add a sort operation T -> S to the logical graph
+         * @param order std::vector<std::size_t> the order of columns (represented
+         * as indices) to sort by
+         * @param orderEnums std::vector<std::size_t> an enum corresponding to
```
Comment (LeonhardFS) [inline tuplex/core/include/DataSet.h:93]: can we change size_t to an enum class which specifies the options explicitly?

Comment (LeonhardFS) [reply tuplex/core/include/DataSet.h:93]: also, consider passing via `const std::vector&` instead of copy.

File: tuplex/core/include/ee/local/LocalBackend.h [hunk 2]
```diff
@@ -50,6 +51,8 @@ namespace tuplex {
         std::vector<Executor*> _executors; //! drivers to be used
         std::unique_ptr<JITCompiler> _compiler;
 
+        void executeSortStage(SortStage* sstage);
```
Comment (LeonhardFS) [inline tuplex/core/include/ee/local/LocalBackend.h:54]: add doxygen comment on what this does (i.e. some details on predecessor stages etc.)

File: tuplex/core/include/logical/SortOperator.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
+//
```
Comment (LeonhardFS) [inline tuplex/core/include/logical/SortOperator.h:1]: add recent Tuplex header

File: tuplex/core/include/logical/SortOperator.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+#ifndef TUPLEX_SORTOPERATOR_H
+#define TUPLEX_SORTOPERATOR_H
+
+#include "LogicalOperator.h"
+
+namespace tuplex {
+    class SortOperator : public LogicalOperator {
+    private:
+        std::vector<size_t> _order;
```
Comment (LeonhardFS) [inline tuplex/core/include/logical/SortOperator.h:13]: rename & add doxygen comments explaining members

File: tuplex/core/include/logical/SortOperator.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        std::vector<size_t> _orderEnum;
+        std::vector<std::string> _columns;
+    public:
+        SortOperator(LogicalOperator *parent, std::vector<size_t> order, std::vector<size_t> orderEnum);
```
Comment (LeonhardFS) [inline tuplex/core/include/logical/SortOperator.h:17]: delete default constructor?

File: tuplex/core/include/logical/SortOperator.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+        std::vector<size_t> orderEnum() { return _orderEnum; }
```
Comment (LeonhardFS) [inline tuplex/core/include/logical/SortOperator.h:19]: make getter functions const (i.e. `std::vector<size_t> indicesToOrderAfter() const { return ... }`

File: tuplex/core/include/logical/SortOperator.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+        std::vector<size_t> order() { return _order; }
+
+        virtual std::string name() override { return "sort"; }
```
Comment (LeonhardFS) [inline tuplex/core/include/logical/SortOperator.h:23]: should be const too? if not, probably should change across all operators (if so, please do in separate commit so it's clear this change has been done when squashing the PR).

File: tuplex/core/include/physical/SortStage.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace tuplex {
+    // forward declaration of friend class
+    namespace codegen {
+        class StageBuilder;
+    }
+
+    /*!
+     * describes a stage which sorts data.
+     * execute as narrow stage with per-node/per-thread parallelism.
```
Comment (LeonhardFS) [inline tuplex/core/include/physical/SortStage.h:18]: can you detail this a bit more? I like that an explanation on what this does is added here!

Comment (LeonhardFS) [reply tuplex/core/include/physical/SortStage.h:18]: because input/output mode is memory, maybe add that a sorting stage requires ALL data to be in main-memory. (I.e., its not able to directly perform external sort)

File: tuplex/core/src/DataSet.cc [hunk 9]
```diff
@@ -163,6 +164,47 @@ namespace tuplex {
...
+    DataSet &DataSet::sort(std::vector<std::size_t> order, std::vector<std::size_t> orderEnums) {
+        // if error dataset, return itself
+        if (isError())
+            return *this;
+
+        assert(_context);
+        assert(this->_operator);
+
+        // parameter checking
+        assert(order.size() == orderEnums.size());
```
Comment (LeonhardFS) [inline tuplex/core/src/DataSet.cc:176]: maybe do this check also in release mode, i.e. explicit if with logger().error()... I fear users might pass wrongly params. Maybe there's a slightly more robust version imaginable? I.e., indices for which orderEnum is not specified?

File: tuplex/core/src/DataSet.cc [hunk 10]
```diff
@@ -163,6 +164,47 @@ namespace tuplex {
...
+
+        // sanity check/print for order and orderEnums
```
Comment (LeonhardFS) [inline tuplex/core/src/DataSet.cc:178]: remove commented code, or switch to logger().debug() and surround with #ifndef

File: tuplex/core/src/DataSet.cc [hunk 11]
```diff
@@ -163,6 +164,47 @@ namespace tuplex {
...
+        //            printf("o: %d, e: %d\n", order[i], orderEnums[i]);
+        //        }
+
+        LogicalOperator *op = _context->addOperator(new SortOperator(this->_operator, order, orderEnums));
+
+        if (!op->good()) {
+            Logger::instance().defaultLogger().error("failed to create sort operator");
+            return _context->makeError("failed to add sort operator to logical plan");
+        }
+        // action, perform it!
```
Comment (LeonhardFS) [inline tuplex/core/src/DataSet.cc:189]: sort is not an action, please update comments.

File: tuplex/core/src/TransformStageExceptionReservoir.cc [hunk 12]
```diff
@@ -406,6 +406,10 @@ namespace tuplex {
                 // op is valid
                 assert(op);
                 switch(op->type()) {
+                    case LogicalOperatorType::SORT: {
+                        // TODO: COLBY
```
Comment (LeonhardFS) [inline tuplex/core/src/TransformStageExceptionReservoir.cc:410]: remove TODO, add description why this can be skipped here. Should logic here be added?

File: tuplex/core/src/ee/local/LocalBackend.cc [hunk 13]
```diff
@@ -14,7 +14,7 @@
 #include <physical/ResolveTask.h>
 #include <physical/TransformTask.h>
 #include <physical/SimpleFileWriteTask.h>
-
+#include "Sort.cc"
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/LocalBackend.cc:17]: why include Sort.cc? Sort.h should be sufficient.

File: tuplex/core/src/ee/local/LocalBackend.cc [hunk 14]
```diff
@@ -132,6 +132,11 @@ namespace tuplex {
         auto tstage = dynamic_cast<TransformStage*>(stage);
         if(tstage)
             executeTransformStage(tstage);
+        else if (dynamic_cast<SortStage*>(stage)) {
+            auto stg = dynamic_cast<SortStage*>(stage);
+            stg->setResultSet(stage->resultSet());
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/LocalBackend.cc:137]: maybe try to remove duplicate dynamic_cast mode (can you use directly check on stage?)

File: tuplex/core/src/logical/LogicalPlan.cc [hunk 15]
```diff
@@ -115,6 +115,11 @@ namespace tuplex {
             assert(udfop);
 
             switch(op->type()) {
+                case LogicalOperatorType::SORT: {
+                    // TODO: COLBY
```
Comment (LeonhardFS) [inline tuplex/core/src/logical/LogicalPlan.cc:119]: remove todo, verify what to do here.

File: tuplex/core/src/ee/local/Sort.cc [hunk 16]
```diff
@@ -0,0 +1,605 @@
+//
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:1]: Tuplex header

File: tuplex/core/src/ee/local/Sort.cc [hunk 17]
```diff
@@ -0,0 +1,605 @@
...
+// Created by Colby Anderson on 6/2/21.
+//
+
+#include <PartitionWriter.h>
+#include "Row.h"
+
+// pair<pair<index # into partitions, row # into partition>, pair<beg. offset, end. offset>>
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:8]: all of this should be put into tuplex namespace.

File: tuplex/core/src/ee/local/Sort.cc [hunk 18]
```diff
@@ -0,0 +1,605 @@
...
+using PartitionSortType = std::pair<std::pair<size_t, size_t>, std::pair<size_t, size_t>>;
+
+// gets the given number of bytes allocated for a specific
+// fixed python type
+static size_t fixedLengthTypeSerializationSize(const python::Type& type) {
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:13]: what about tuplex of fixedLength type?

Comment (LeonhardFS) [reply tuplex/core/src/ee/local/Sort.cc:13]: maybe add a test for this function with all the supported types

File: tuplex/core/src/ee/local/Sort.cc [hunk 19]
```diff
@@ -0,0 +1,605 @@
...
+    assert(type.isNumericType());
+    if (type.isNumericType()) {
+        return sizeof(int64_t);
+    } else {
+        throw std::runtime_error("Error: Couldn't recognize fixed type");
+    }
+}
+
+// offsets of the form <offset of beginning of row from start of partition...
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:22]: change doc strings to doxygen comment style.

File: tuplex/core/src/ee/local/Sort.cc [hunk 20]
```diff
@@ -0,0 +1,605 @@
...
+    std::vector<size_t> ret;
+    if (order.empty()) {
+        for (int i = 0; i < numColumns; i++) {
+            ret.push_back(i);
+        }
+    } else {
+        return order;
+    }
+    return ret;
+    //    for (int i = 0; i < numColumns; i++) {
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:94]: remove comments.

File: tuplex/core/src/ee/local/Sort.cc [hunk 21]
```diff
@@ -0,0 +1,605 @@
...
+    if (orderEnum.empty()) {
+        for (int i = 0; i < numColumns; i++) {
+            ret.push_back(i);
+        }
+    } else {
+        return orderEnum;
+    }
+    return ret;
+    //
+    //    for (int i = 0; i < numColumns; i++) {
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:114]: remove commented. Add doxygen comments to all functions. Consider passing vectors via `const vector<...>&`

File: tuplex/core/src/ee/local/Sort.cc [hunk 22]
```diff
@@ -0,0 +1,605 @@
...
+            int currentColIndex = order.at(counter);
+            // gets the current type of this current column
+            python::Type colType = colTypes.at(currentColIndex);
+            // sorting for integer
+            counter++;
+            if (colType == python::Type::I64) {
+                int64_t lInt = lRow.getInt(currentColIndex);
+                int64_t rInt = rRow.getInt(currentColIndex);
+                if (lInt == rInt) {continue;} // edge case where there is no other column to sort by
+                else if (orderEnum.at(counter-1) == 3) {
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:176]: do not use fixed numbers, rather enum class. Magic numbers are hard to debug.

File: tuplex/core/src/ee/local/Sort.cc [hunk 23]
```diff
@@ -0,0 +1,605 @@
...
+                    std::ostringstream o1;
+                    std::ostringstream o2;
+                    o1 << lInt;
+                    o2 << rInt;
+                    std::string lStrInt = o1.str();
+                    std::string rStrInt = o2.str();
+                    return lStrInt.length() < rStrInt.length();
+                }
+                else if (orderEnum.at(counter-1) == 4) {
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:185]: maybe intro a helper variable storing result of counter - 1.

Comment (LeonhardFS) [reply tuplex/core/src/ee/local/Sort.cc:185]: also why does counter first get inced, and then when it's used always decreased by 1?

File: tuplex/core/src/ee/local/Sort.cc [hunk 24]
```diff
@@ -0,0 +1,605 @@
...
+        // gets the "left" row
+        tuplex::Row lRow = tuplex::Row::fromMemory(schema, partitionPtrs.at(l.first.first) + l.second.first,
+                                                   l.second.second - l.second.first);
+        // gets the "right" row
+        tuplex::Row rRow = tuplex::Row::fromMemory(schema, partitionPtrs.at(r.first.first) + r.second.first,
+                                                   r.second.second - r.second.first);
+        // gets a list of the types for each column
+        std::vector<python::Type> colTypes = schema.getRowType().parameters();
+        int counter = 0;
+        while (counter != order.size()) {
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:165]: why not a for loop?

File: tuplex/core/src/ee/local/Sort.cc [hunk 25]
```diff
@@ -0,0 +1,605 @@
...
+                }
+                else if (orderEnum.at(counter-1) == 2) {
+                    return lInt > rInt;
+                }
+                else {
+                    return lInt < rInt;
+                }
+                // sorting for floats
+            } else if (colType == python::Type::F64) {
+                double lDub = lRow.getDouble(currentColIndex);
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:220]: maybe add note. Comparing string size with float formats could lead to funny results. Let's try to find the right strategy here & describe here.

File: tuplex/core/src/ee/local/Sort.cc [hunk 26]
```diff
@@ -0,0 +1,605 @@
...
+                }
+                else if (orderEnum.at(counter-1) == 2) {
+                    return lBool > rBool;
+                }
+                else {
+                    return lBool < rBool;
+                }
+                // doesn't support other types yet
+            }
+            // TODO: Test Handle Generic Tuple
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:328]: remove todos

File: tuplex/core/src/ee/local/Sort.cc [hunk 27]
```diff
@@ -0,0 +1,605 @@
...
+                    std::string rStrBool = o2.str();
+                    return lStrBool > rStrBool;
+                }
+                else if (orderEnum.at(counter-1) == 2) {
+                    return lBool > rBool;
+                }
+                else {
+                    return lBool < rBool;
+                }
+                // doesn't support other types yet
```
Comment (LeonhardFS) [inline tuplex/core/src/ee/local/Sort.cc:326]: error message for this?

File: tuplex/core/include/DataSet.h (hunk 1 above)
Comment (colby-anderson) [reply tuplex/core/include/DataSet.h:93]: Done.

File: tuplex/core/include/ee/local/LocalBackend.h (hunk 2 above)
Comment (colby-anderson) [reply tuplex/core/include/ee/local/LocalBackend.h:54]: Done.

File: tuplex/core/include/logical/SortOperator.h (hunk 3 above)
Comment (colby-anderson) [reply tuplex/core/include/logical/SortOperator.h:1]: Done.

File: tuplex/test/wrappers/WrapperTest.cc [hunk 28]
```diff
@@ -1552,6 +1552,25 @@ TEST_F(WrapperTest, TupleParallelizeI) {
     }
 }
 
+TEST_F(WrapperTest, Sort) {
+    using namespace tuplex;
+
+    PyObject* listObj = python::runAndGet("L = [5, 3, 1, 4, 2]", "L");
```
Comment (LeonhardFS) [inline tuplex/test/wrappers/WrapperTest.cc:1558]: can we expand this test a bit? I.e. sorting after different columns etc.? I'd also like to have one end-to-end test which sorts all the csv files after each column, and let's say combinations of 2-3 columns. Just to be sure the sorting feature works 100%.

Comment (LeonhardFS) [reply tuplex/test/wrappers/WrapperTest.cc:1558]: a good idea might be also, to vary the partitionSize for a large file, to be sure overlapping partitions are correctly handled.

File: tuplex/python/tests/test_sort.py [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+        ds = c.parallelize(["az", "baaz", "aaa"])
+        d = {0:SortBy.DESCENDING_LENGTH}
+        res1 = ds.sort(by=d)
+        print("D", res1.collect())
+        assert res1.collect() == ["baaz", "aaa", "az"]
+
+        # no duplicates. 1 col. length cast asc
+        c = Context(self.conf)
+        ds = c.parallelize([19, 111, 5])
+        d = {0:SortBy.ASCENDING_LENGTH}
```
Comment (LeonhardFS) [inline tuplex/python/tests/test_sort.py:55]: do we support key lookup via column names as well?

File: tuplex/python/src/PythonDataSet.cc [hunk 30]
```diff
@@ -105,6 +105,72 @@ namespace tuplex {
...
+            PyObject * obj2 = PyList_GetItem(listObj2, i);
+
+            // check type:
+            if (PyObject_IsInstance(obj1, reinterpret_cast<PyObject *>(&PyLong_Type)) &&
+                PyObject_IsInstance(obj2, reinterpret_cast<PyObject *>(&PyLong_Type))) {
+                int index = boost::python::extract<int>(order[i]);
+                int enumSort = boost::python::extract<int>(enumInOrder[i]);
+
+                if (index < 0 || index >= num_cols || enumSort <= 0 || enumSort > 6) {
+                    throw std::runtime_error("invalid index " + std::to_string(index)
```
Comment (LeonhardFS) [inline tuplex/python/src/PythonDataSet.cc:150]: does this work? I.e., can we have a test which violates that constraint? (in py.test)

File: tuplex/python/include/PythonDataSet.h [hunk 31]
```diff
@@ -49,6 +49,8 @@ namespace tuplex {
 
         PythonDataSet unique();
 
+        PythonDataSet sort(boost::python::list order, boost::python::list enumInOrder);
```
Comment (LeonhardFS) [inline tuplex/python/include/PythonDataSet.h:52]: pls add doxygen string what the formats of the lists should be like.

File: tuplex/core/src/physical/StageBuilder.cc [hunk 32]
```diff
@@ -676,6 +679,10 @@ namespace tuplex {
                 assert(node);
                 UDFOperator *udfop = dynamic_cast<UDFOperator *>(node);
                 switch (node->type()) {
+                    case LogicalOperatorType::SORT: {
+                        // TODO: COLBY
```
Comment (LeonhardFS) [inline tuplex/core/src/physical/StageBuilder.cc:683]: can we resolve the TODO here?

File: tuplex/core/src/physical/SortStage.cc [hunk 33]
```diff
@@ -0,0 +1,36 @@
...
+        for(auto stage: predecessors())
+            ss<<"Stage"<<stage->number()<<" ";
+        Logger::instance().defaultLogger().info(ss.str());
+
+        // execute all predecessors (can be at most one!)
+        // @TODO: this should be parallelized for tiny stages!
+        for (auto stage : predecessors())
+            stage->execute(context);
+
+        //        auto stage = predecessors()[0];
```
Comment (LeonhardFS) [inline tuplex/core/src/physical/SortStage.cc:30]: remove commented code, also what's up with the auto x = this? why not directly use backend()->execute(this) ?

**Review by LeonhardFS — CHANGES_REQUESTED** [review]
pls address the comments, thx!
